src/tex/crunch_functions.cpp

```diff
@@ -0,0 +1,110 @@
+#include <BS_thread_pool.hpp>
+#include <btu/tex/crunch_functions.hpp>
+#include <btu/tex/formats.hpp>
+#include <crunch/crnlib.h>
+
+#include <algorithm>
+
+namespace btu::tex {
+using crnlib::dxt_image;
+using crnlib::pixel_format;
+
+auto resize(CrunchTexture &&file, Dimension dim) -> ResultCrunch
+{
+    // Resizes the input texture. If compressed, automatically decompresses it. Removes mipmaps.
+    mipmapped_texture::resample_params res_params;
+    res_params.m_filter_scale  = 1.0f;
+    res_params.m_multithreaded = true;
+
+    // Diffuse textures are assumed to be sRGB.
+    if (file.get_texture_type() == TextureType::Diffuse)
+    {
+        res_params.m_gamma = 2.2f;
+        res_params.m_srgb  = true;
+    }
+    else
+    {
+        res_params.m_gamma = 1.0f;
+        res_params.m_srgb  = false;
+    }
+
+    const auto success = file.get().resize(dim.w, dim.h, res_params);
+    if (!success)
+    {
+        // TODO: Better error propagation.
+        return tl::make_unexpected(error_from_hresult(E_UNEXPECTED));
+    }
+
+    return std::move(file);
+}
+
+auto generate_mipmaps(CrunchTexture &&file) -> ResultCrunch
+{
+    mipmapped_texture::generate_mipmap_params gen_params;
+    gen_params.m_multithreaded = true;
+    gen_params.m_max_mips      = crn_limits::cCRNMaxLevels;
+    gen_params.m_min_mip_size  = 1;
+
+    if (file.get_texture_type() == TextureType::Diffuse)
+    {
+        gen_params.m_gamma = 2.2f;
+        gen_params.m_srgb  = true;
+    }
+    else
+    {
+        gen_params.m_gamma = 1.0f;
+        gen_params.m_srgb  = false;
+    }
+
+    const auto success = file.get().generate_mipmaps(gen_params, true);
+    if (!success)
+    {
+        // TODO: Better error propagation.
+        return tl::make_unexpected(error_from_hresult(E_UNEXPECTED));
+    }
+
+    return std::move(file);
+}
+
+auto convert(CrunchTexture &&file, DXGI_FORMAT format) -> ResultCrunch
+{
+    pixel_format crunch_format;
+    switch (format)
+    {
+        case DXGI_FORMAT_BC1_UNORM: crunch_format = pixel_format::PIXEL_FMT_DXT1; break;
+        case DXGI_FORMAT_BC3_UNORM: crunch_format = pixel_format::PIXEL_FMT_DXT5; break;
+        case DXGI_FORMAT_B8G8R8X8_UNORM: crunch_format = pixel_format::PIXEL_FMT_R8G8B8; break;
+        case DXGI_FORMAT_B8G8R8A8_UNORM:
+        case DXGI_FORMAT_R8G8B8A8_UNORM: crunch_format = pixel_format::PIXEL_FMT_A8R8G8B8; break;
+        default: return tl::make_unexpected(error_from_hresult(E_UNEXPECTED));
+    }
+
+    // Default compression arguments.
+    dxt_image::pack_params pack_params;
+
+    // TODO: Use threads? How many? Quite slow without them.
```

Comment:
For now, let's keep it like this. But I'm really starting to wonder if bethutil multithreading is useful, as most underlying operations are already multithreaded

Reply:
Yeah, I have a feeling the bethutil multithreading is not really helping that much. Could be tested out with standard DirectX, see if it is in fact speeding the processing up or not (by setting concurrency to 1).